src/number_theory/arithmetic_function.lean

```diff
@@ -221,15 +232,317 @@ instance : semiring (arithmetic_function α) :=
+    { transitivity nat.gcd (b1 * b2) (a2 * b2),
+      {rw [nat.gcd_mul_right,
+           cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.2.1, h.2, nat.gcd_mul_right,
+            cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul] } } },
+  { rintros ⟨b1, b2⟩ h,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at h,
+    use ((b1.gcd m, b2.gcd m), (b1.gcd n, b2.gcd n)),
+    simp only [exists_prop, prod.mk.inj_iff, ne.def, mem_product, mem_divisors_antidiagonal],
+    rw [← cop.gcd_mul _, ← cop.gcd_mul _, ← h.1, nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop h.1,
+        nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop.symm _],
+    { rw [nat.mul_eq_zero, decidable.not_or_iff_and_not] at h, simp [h.2.1, h.2.2] },
+    rw [mul_comm n m, h.1] }
+end⟩
+
+lemma pmul [comm_semiring R] {f g : arithmetic_function R}
+  (hf : f.is_multiplicative)  (hg : g.is_multiplicative) :
+  is_multiplicative (f.pmul g) :=
+⟨by { simp [hf, hg], }, λ m n cop, begin
+  simp only [pmul_apply, hf.map_mul_of_coprime cop, hg.map_mul_of_coprime cop],
+  ring,
+end⟩
+
+end is_multiplicative
+
+section special_functions
+
+/-- The identity on `ℕ` as an `arithmetic_function`.  -/
+def id : arithmetic_function ℕ := ⟨id, rfl⟩
+
+@[simp]
+lemma id_apply {x : ℕ} : id x = x := rfl
+
+/-- `pow k n = n ^ k`, except `pow 0 0 = 0`. -/
+def pow (k : ℕ) : arithmetic_function ℕ := id.ppow k
+
+@[simp]
+lemma pow_apply {k n : ℕ} : pow k n = if (k = 0 ∧ n = 0) then 0 else n ^ k :=
+begin
+  cases k,
+  { simp [pow] },
+  simp [pow, (ne_of_lt (nat.succ_pos k)).symm],
+end
+
+/-- `σ k n` is the sum of the `k`th powers of the divisors of `n` -/
+def sigma (k : ℕ) : arithmetic_function ℕ :=
+⟨λ n, ∑ d in divisors n, d ^ k, by simp⟩
+
+localized "notation `σ` := sigma" in arithmetic_function
+
+@[simp]
+lemma sigma_apply {k n : ℕ} : σ k n = ∑ d in divisors n, d ^ k := rfl
+
+lemma zeta_mul_pow_eq_sigma {k : ℕ} : ζ * pow k = σ k :=
+begin
+  ext,
+  rw [sigma, zeta_mul_apply],
+  apply sum_congr rfl,
+  intros x hx,
+  rw [pow_apply, if_neg (not_and_of_not_right _ _)],
+  contrapose! hx,
+  simp [hx],
+end
+
+lemma is_multiplicative_zeta [semiring R] : is_multiplicative (ζ : arithmetic_function R) :=
+⟨by simp, λ m n cop, begin
+  cases m, {simp},
+  cases n, {simp},
+  rw [zeta_apply_ne (mul_ne_zero _ _), zeta_apply_ne,
+      zeta_apply_ne, mul_one],
+  repeat { apply nat.succ_ne_zero },
+end⟩
+
+lemma is_multiplicative_id : is_multiplicative arithmetic_function.id :=
+⟨rfl, λ _ _ _, rfl⟩
+
+lemma is_multiplicative.ppow [comm_semiring R] {f : arithmetic_function R}
+  (hf : f.is_multiplicative) {k : ℕ} :
+  is_multiplicative (f.ppow k) :=
+begin
+  induction k with k hi,
+  { exact is_multiplicative_zeta },
+  { rw ppow_succ,
+    apply hf.pmul hi },
+end
+
+lemma is_multiplicative_pmul {k : ℕ} : is_multiplicative (pow k) :=
```

Comment:
This is a duplicate of the lemma below it: did you intend a different statement?